lib/log_ratelimit.c

```diff
@@ -0,0 +1,90 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <rte_cycles.h>
+#include <rte_lcore.h>
+
+#include "gatekeeper_main.h"
+#include "gatekeeper_log_ratelimit.h"
+
+static struct log_ratelimit_state log_ratelimit_states[RTE_MAX_LCORE];
+
+void
+log_ratelimit_state_init(unsigned lcore_id,
+	uint32_t ratelimit_level, uint32_t interval, uint32_t burst)
+{
+	struct log_ratelimit_state *lrs;
+
+	RTE_VERIFY(lcore_id < RTE_MAX_LCORE);
+
+	lrs = &log_ratelimit_states[lcore_id];
+	lrs->ratelimit_level = ratelimit_level;
+	lrs->interval_cycles = interval * cycles_per_ms;
+	lrs->burst = burst;
+	lrs->printed = 0;
+	lrs->begin = rte_rdtsc();
+}
+
+/*
+ * Rate limiting log entries.
+ *
+ * Returns:
+ * - true means go ahead and do it.
+ * - false means callbacks will be suppressed.
+ */
+bool
+log_ratelimit_allow(struct log_ratelimit_state *lrs)
```

Comment:
`log_ratelimit_allow()` is only used in `rte_log_ratelimit()`, shouldn't it be `static`?